pkg/controller/jobframework/reconciler.go

```diff
@@ -569,34 +569,60 @@ func (r *JobReconciler) recordAdmissionCheckUpdate(wl *kueue.Workload, job Gener
 	}
 }
 
-// IsParentJobManaged checks whether the parent job is managed by kueue.
-func (r *JobReconciler) IsParentJobManaged(ctx context.Context, jobObj client.Object, namespace string) (bool, error) {
-	owner := metav1.GetControllerOf(jobObj)
-	if owner == nil {
-		return false, nil
-	}
-	parentJob := GetEmptyOwnerObject(owner)
-	if parentJob == nil {
-		return false, fmt.Errorf("workload owner %v: %w", owner, ErrUnknownWorkloadOwner)
-	}
-	if err := r.client.Get(ctx, client.ObjectKey{Name: owner.Name, Namespace: namespace}, parentJob); err != nil {
-		return false, errors.Join(ErrWorkloadOwnerNotFound, err)
+// IsAncestorJobManaged checks whether an ancestor job is managed by kueue.
+func (r *JobReconciler) IsAncestorJobManaged(ctx context.Context, jobObj client.Object, namespace string) (bool, error) {
+	ancestor, err := r.getAncestorJobManagedByKueue(ctx, jobObj, namespace)
+	if err != nil {
+		return false, err
 	}
-	return QueueNameForObject(parentJob) != "", nil
+	return ancestor != nil, nil
 }
 
-func (r *JobReconciler) getParentWorkload(ctx context.Context, object client.Object) (*kueue.Workload, error) {
-	owner := metav1.GetControllerOf(object)
-	if owner == nil {
+// getAncestorWorkload returns the Workload object of the Kueue-managed ancestor job.
+func (r *JobReconciler) getAncestorWorkload(ctx context.Context, jobObj client.Object, namespace string) (*kueue.Workload, error) {
+	ancestor, err := r.getAncestorJobManagedByKueue(ctx, jobObj, namespace)
+	if err != nil {
+		return nil, err
+	}
+	if ancestor == nil {
 		return nil, nil
 	}
 	wlList := kueue.WorkloadList{}
-	if err := r.client.List(ctx, &wlList, client.InNamespace(object.GetNamespace()), client.MatchingFields{indexer.OwnerReferenceUID: string(owner.UID)}); err != nil || len(wlList.Items) == 0 {
-		return nil, client.IgnoreNotFound(err)
+	if err := r.client.List(ctx, &wlList, client.InNamespace(ancestor.GetNamespace()), client.MatchingFields{indexer.OwnerReferenceUID: string(ancestor.GetUID())}); client.IgnoreNotFound(err) != nil {
+		return nil, err
+	}
+	if len(wlList.Items) > 0 {
+		// In theory the job can own multiple Workloads, we cannot do too much about it, maybe log it.
+		return &wlList.Items[0], nil
 	}
+	return nil, nil
+}
 
-	// In theory the parent can own multiple Workloads, we cannot do too much about it, maybe log it.
-	return &wlList.Items[0], nil
+// getAncestorJobManagedByKueue traverses controllerRefs to find an ancestor job that is manged by Kueue (ie, it has a queue-name label).
+func (r *JobReconciler) getAncestorJobManagedByKueue(ctx context.Context, jobObj client.Object, namespace string) (client.Object, error) {
+	seen := make(sets.Set[types.UID])
```

Comment:
```suggestion
	seen := sets.New[types.UID]()
```

nit: Actually, we have `New` function to initialize sets.

Reply:
thanks.  folded into #4300.